src/pluginextension.h

```diff
@@ -0,0 +1,76 @@
+/*
+ *      pluginextension.h - this file is part of Geany, a fast and lightweight IDE
+ *
+ *      Copyright 2023 The Geany contributors
+ *
+ *      This program is free software; you can redistribute it and/or modify
+ *      it under the terms of the GNU General Public License as published by
+ *      the Free Software Foundation; either version 2 of the License, or
+ *      (at your option) any later version.
+ *
+ *      This program is distributed in the hope that it will be useful,
+ *      but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *      MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *      GNU General Public License for more details.
+ *
+ *      You should have received a copy of the GNU General Public License along
+ *      with this program; if not, write to the Free Software Foundation, Inc.,
+ *      51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#ifndef GEANY_PLUGIN_EXTENSION_H
+#define GEANY_PLUGIN_EXTENSION_H 1
+
+#include "document.h"
+
+/* Temporary define so the plugin can check whether it's compiled against
+ * Geany with plugin extension support or not */
+#define HAVE_GEANY_PLUGIN_EXTENSION 1
+
+G_BEGIN_DECLS
+
+
+typedef struct {
+	gboolean (*autocomplete_provided)(GeanyDocument *doc);
+	void (*autocomplete_perform)(GeanyDocument *doc, gboolean force);
+
+	gboolean (*calltips_provided)(GeanyDocument *doc);
+	void (*calltips_show)(GeanyDocument *doc, gboolean force);
+
+	gboolean (*goto_provided)(GeanyDocument *doc);
+	void (*goto_perform)(GeanyDocument *doc, gint pos, gboolean definition);
+
+	gboolean (*doc_symbols_provided)(GeanyDocument *doc);
+	GPtrArray *(*doc_symbols_get)(GeanyDocument *doc);
+
+	gboolean (*symbol_highlight_provided)(GeanyDocument *doc);
+
+	gchar _dummy[1024];
+} PluginExtension;
+
+
+void plugin_extension_register(PluginExtension *extension);
+void plugin_extension_unregister(PluginExtension *extension);
```

Comment:
Ah also: do we want to add user data to those functions?  That could be handy to pass in the general plugin data, especially as the "new" (which isn't so new anymore) plugin API allows this.

Reply:
OK, I may not understand this completely, but the idea is that the decision on whether the extension performs the particular feature is directed by the extension itself (so it decides whether it's "active" or not). This is basically what I described in https://github.com/geany/geany/pull/3849#discussion_r1625127493: the return value of the `plugin_extension_goto_provided(doc)` function decides whether the extension wants to perform the feature and the `plugin_extension_goto_perform(doc)` function performs it. The LSP plugin for instance takes a look at the filetype of `doc` and when there is a server configured for it and in addition, this particular server supports goto definition/declaration (this is obtained from the initial handshake with the server), `plugin_extension_goto_provided(doc)` returns TRUE, otherwise it returns FALSE. In the TRUE case, `plugin_extension_goto_perform(doc)` can be called.

Reply:
(just a few quick replies, I'll go deeper later)

My point is that currently there is no way for a extension to know whether it "won" or not, so it can't know whether it should do ancillary actions (e.g. things in editor-notify, see e.g. my diff) related to the feature or not (assuming not everything does in `_perform()`).
What I am trying to suggest is that there should be a means for the extension to know that, instead of having a setting the user need to know about and tweak (also knowing which plugin effectively "won").
*Maybe* it's fine and the no two extension can return `TRUE` for the same `provided()` call, but it seems a bit brittle and hard to get to work long-term.

A "simple" (but not very elegant I'd say) solution could be *always* calling `_provided()` at least once per document, and the extensions *must* track whether they got called for a document (and so it didn't got "caught" by another extension) before doing anything else.

Basically my point is that if we wanted to have the Geany features use the extension mechanism, how would it know whether it's active or not?  It's a fallback for *all* calls (basically returning `TRUE` for all the `_provided()` calls), and so the question is "was it overridden"?

This sort of support could be useful for other things if we expand it further, e.g. a fairly generic indenter based on a fairly generic algorithm, and a specialized one (e.g. clang-format) that both handle some of the same documents, but with different levels of "quality".  Currently I don't see a solution for those 2 to work together, unless each has an option to select what it supports or not, and it's properly configured (which could be a nightmare).

Reply:
OK, what about modifying every `_provided(GeanyDocument *doc)` to 
```
_provided(GeanyDocument *doc, PluginExtension *ext)
```
which:
* if `ext != NULL`, returns TRUE only if the passed `ext` is the one that won, FALSE otherwise, 
* if `ext == NULL` it behaves like now?

(And, in addition, maybe `_active()` is a better naming than `_provided()`)

Reply:
> OK, what about modifying every `_provided(GeanyDocument *doc)` to […]

Yeah that's a good idea, and I came to it as well partly separately, so it might be a good, or at least obvious, answer :)

> Where would this `user_data` get passed? When you call `_perform(..., user_data)`, the plugin won't know what to do with the data as it comes from Geany and it doesn't have any callback where it would pass it. The `_provided()` functions are synchronous and also don't know what to do with such data.

I guess we didn't understand each other here.  My point is that at `_register()` time the extension gives some data, and when calling the vfuncs it gets passed in, allowing to remove all global state in the extension's implementation.
E.g. for a C++ wrapper it could do (pardon my rusty C++):

```c++
void foobar_provided(GeanyDocument *doc, gpointer data)
{
    static_cast<Foo*>(data)->foobar_provided_impl(doc);
}

Foo::Foo()
{
    plugin_extension_register(vtable, this);
}
```

Reply:
> My point is that at _register() time...

Alright, it was the `_register()` function and I get the example you gave. But I don't understand what global state you mean in the PluginExtension case. When I convert all the `_provided()` and `_perform()` functions to go through the list of registered extensions, we still need to store this list "globally" in the file. And there won't be any other global variable, the "dummy" LSP will disappear.

Reply:
OK, I just managed to scroll to your patch and the data is meant to be used by Geany (or whatever wants to implement the plugin) and not the plugin or the API which is what I mistunderstood.

Reply:
> OK, I just managed to scroll to your patch and the data is meant to be used by Geany (or whatever wants to implement the plugin) and not the plugin or the API which is what I mistunderstood.

Midnight again, I stop making sense. Anyway, I got it :-).